src/Microsoft.IdentityModel.Tokens/Cache/MaximumSizeCache.cs

```diff
@@ -0,0 +1,272 @@
+                value = val.Value;
+
+            return found;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            value = default;
+
+            var found = TryRemoveInternal(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val.Value;
+
+            return found;
+        }
+
+        #endregion
+
+
+        #region protected
+
+        /// <summary>
+        /// The internal version of TryGetValue that gets the LRUCacheItem from the hash.
+        /// </summary>
+        /// <param name="key">The key of the value to get.</param>
+        /// <param name="value">The object that has the specified key, or the default value of the type if the not found.</param>
+        /// <returns>true if the key was found, otherwise, false.</returns>
+        protected virtual bool TryGetValueInternal(TKey key, out CacheItem<TKey, TValue> value)
+        {
+            value = default;
+
+            var found = _map.TryGetValue(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val;
+
+            return found;
+        }
+
+        /// <summary>
+        /// Attempts to remove and return the value that has the specified key from the cache.
+        /// </summary>
+        /// <param name="key">The key of the element to remove and return.</param>
+        /// <param name="value">The object removed from the cache, or the default value of the TValue type if key does not exist.</param>
+        /// <returns>true if the object was removed successfully; otherwise, false.</returns>
+        protected virtual bool TryRemoveInternal(TKey key, out CacheItem<TKey, TValue> value)
+        {
+            value = default;
+
+            var found = _map.TryRemove(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val;
+
+            return found;
+        }
+
+        /// <summary>
+        /// Validate the key, value and check the expiration time before adding the item is added to the cache.
+        /// </summary>
+        /// <param name="key">The key of the item to be added to the cache.</param>
+        /// <param name="value">The value of the item to be added to the cache.</param>
+        /// <param name="expirationTime">The expiration time of the item.</param>
+        /// <returns>true if valid, false otherwise.</returns>
+        protected static bool ValidateValues(TKey key, TValue value, DateTime expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            // if item already expired, do not add it to the cache
+            return expirationTime >= DateTime.UtcNow;
```

Comment:
This cache is names MaximumSizeCache, why the check for expirationTime?
EventBasedLRUCache may have RemovedExpiredItems off this seems contradictory.

Reply:
You are right, only the LRU should track the time if specified. This has been changed.